apps/emqx_retainer/src/emqx_retainer_mnesia.erl

```diff
@@ -91,9 +91,10 @@ store_retained(_, Msg =#message{topic = Topic}) ->
                                                          expiry_time = ExpiryTime},
                                                write);
                               [] ->
-                                  ?LOG(error,
-                                       "Cannot retain message(topic=~ts) for table is full!",
-                                       [Topic]),
+                                  ?SLOG(error,
```

Comment:
We should not write logs in mnesia transaction!
abort the transaction here with
`mnesia:abort(table_is_full)`
and later:

```
case mria:transaction(?RETAINER_SHARD, Fun) of
    {atomic, ok} ->  ok;
    {aborted, Reason} -> ?SLOG(error, #{msg => "failed_to_retain_message", topic => Topic, reason => Reason})
end
```